etw/eventmetadata.go

```diff
@@ -0,0 +1,51 @@
+package etw
+
+import (
+	"bytes"
+	"encoding/binary"
+)
+
+// InType indicates the type of data contained in the ETW event.
+type InType byte
+
+// Various InType definitions for tracelogging.
+const (
+	InTypeNull InType = iota
+	InTypeUnicodeString
+	InTypeAnsiString
+	InTypeInt8
+	InTypeUint8
+	InTypeInt16
+	InTypeUint16
+	InTypeInt32
+	InTypeUint32
+	InTypeInt64
+	InTypeUint64
+	InTypeFloat
+	InTypeDouble
+	InTypeBool32
+)
+
+// EventMetadata maintains a buffer which builds up the metadatadata for an ETW
+// event. It needs to be paired with EventData which describes the event.
+type EventMetadata struct {
+	buffer bytes.Buffer
+}
+
+// NewEventMetadata returns a new EventMetadata with event name and initial
+// metadata written to the buffer.
+func NewEventMetadata(name string) *EventMetadata {
+	em := EventMetadata{}
+	binary.Write(&em.buffer, binary.LittleEndian, uint16(0))    // Length placeholder

```

Comment:
Do you really need the `&em.buffer` seems like since `bytes.Buffer` is `io.Writer` you should be able to just pass `em.buffer`.

Reply:
I tried that, but Go doesn't like it:

```
cannot use em.buffer (type bytes.Buffer) as type io.Writer in argument to binary.Write:
        bytes.Buffer does not implement io.Writer (Write method has pointer receiver)
```

Sounds like there's a `func (*bytes.Buffer) Write` but not a `func (bytes.Buffer) Write`?

Reply:
Oh I see. No worries then. The compiler in my head made a mistake :)